src/Platform/Microsoft.Testing.Platform/OutputDevice/Console/ConsoleLogger.cs

```diff
@@ -0,0 +1,742 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+#if !NET7_0_OR_GREATER
+using System.Diagnostics.CodeAnalysis;
+#endif
+using System.Globalization;
+#if !NET7_0_OR_GREATER
+using System.Reflection;
+
+#endif
+
+using System.Text.RegularExpressions;
+
+using Microsoft.Testing.Platform.Helpers;
+using Microsoft.Testing.Platform.Resources;
+
+namespace Microsoft.Testing.Platform.OutputDevice.Console;
+
+/// <summary>
+/// Console logger that outputs test progress and is capable of writing ANSI or non-ANSI output via the given terminal.
+/// </summary>
+internal partial class ConsoleLogger : IDisposable
+{
+    private static readonly string[] NewLineStrings = { "\r\n", "\n" };
+
+    internal const string SingleIndentation = "  ";
+
+    internal const string DoubleIndentation = $"{SingleIndentation}{SingleIndentation}";
+
+    internal const string TripleIndentation = $"{SingleIndentation}{SingleIndentation}{SingleIndentation}";
+
+    internal Func<StopwatchAbstraction> CreateStopwatch { get; set; } = SystemStopwatch.StartNew;
+
+    /// <summary>
+    /// The two directory separator characters to be passed to methods like <see cref="string.IndexOfAny(char[])"/>.
+    /// </summary>
+    private readonly Dictionary<string, TestModule> _assemblies = new();
```

Comment:
Shall we handle parallelism?

Reply:
No the invocations are coming from the message bus which is serial.